### PR TITLE
fix: show delete action for labels

### LINKED ISSUE #2163: Remove button is not displayed for external labels
### Describe the Bug

image

### Steps to Reproduce

1. Select an external label
2. Only color picker is available

### Expected Behavior

The button should be displayed.

### Environment

 - Browser: [e.g. IE 11, Chrome 69]
 - OS: [e.g. Windows 7]
 - Library version: [e.g. 2.0.0]

---

Initially reported via https://github.com/camunda/improved-canvas/issues/66

### PR DESCRIPTION
**Pull request by jarekdanielak**
Closes #2163

### Proposed Changes

Show delete action for element's label.

image

 

### Checklist

To ensure you provided everything we need to look at your PR:

* [X] **Brief textual description** of the changes present
* [X] **Visual demo** attached
* [ ] **Steps to try out** present, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)
* [X] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

### CODE REVIEW

File: lib/features/context-pad/ContextPadProvider.js [hunk 1]
```diff
@@ -172,9 +172,7 @@ ContextPadProvider.prototype._isDeleteAllowed = function(elements) {
   });
 
   if (isArray(baseAllowed)) {
-    return every(baseAllowed, function(element) {
-      return includes(baseAllowed, element);
```
Comment (barmac) [inline lib/features/context-pad/ContextPadProvider.js:176]: How come this had not been noticed for 5 years 🤡

File: lib/features/context-pad/ContextPadProvider.js [hunk 2]
```diff
@@ -192,14 +190,26 @@ ContextPadProvider.prototype.getContextPadEntries = function(element) {
...
-      rules = this._rules,
       autoPlace = this._autoPlace,
       translate = this._translate,
       appendPreview = this._appendPreview;
 
   var actions = {};
 
   if (element.type === 'label') {
+    if (this._isDeleteAllowed([ element ])) {
+      assign(actions, {
```
Comment (nikku) [inline lib/features/context-pad/ContextPadProvider.js:201]: Could be de-duplicated (cf. delete action later on :arrow_down:).

File: lib/features/context-pad/ContextPadProvider.js [hunk 3]
```diff
@@ -537,16 +547,7 @@ ContextPadProvider.prototype.getContextPadEntries = function(element) {
...
-  var deleteAllowed = rules.allowed('elements.delete', { elements: [ element ] });
-
-  if (isArray(deleteAllowed)) {
-
-    // was the element returned as a deletion candidate?
-    deleteAllowed = deleteAllowed[0] === element;
-  }
-
-  if (deleteAllowed) {
+  if (this._isDeleteAllowed([ element ])) {
```
Comment (nikku) [inline lib/features/context-pad/ContextPadProvider.js:550]: :+1: for removing this duplication.

**Review by nikku — COMMENTED** [review]
Potential to further de-duplicate here (make `delete` a re-usable action).

**Comment by lmbateman** [comment]
Just to clarify, does the icon only show up when the label is in edit mode? And does it show up on all elements? Thanks!

**Comment by nikku** [comment]
It now consistently shows up for _all_ elements, in the editor ("edit mode"). We could decide to (1) hide it for all elements, (2) show it for all elements or (3) keep the (perceived as broken by @barmac) state of not showing it for labels.

**Comment by barmac** [comment]
The reasoning: I can remove the label via keyboard shortcut (backspace/delete), I can remove it via direct editing, so I should be able to remove it with mouse via context pad.
If a modeler implementor disallows label removal via rules, the icon won't show up.

> Just to clarify, does the icon only show up when the label is in edit mode? And does it show up on all elements? Thanks!

The bug report comes from the notion that I can delete the element from the context pad, but I cannot delete the label. However, if I remove the element, its label is removed as well.

File: lib/features/context-pad/ContextPadProvider.js (hunk 2 above)
Comment (jarekdanielak) [reply lib/features/context-pad/ContextPadProvider.js]: 551bba6 extracts adding delete action to a helper function.

File: lib/features/context-pad/ContextPadProvider.js [hunk 4]
```diff
@@ -192,14 +190,17 @@ ContextPadProvider.prototype.getContextPadEntries = function(element) {
...
-      rules = this._rules,
       autoPlace = this._autoPlace,
       translate = this._translate,
       appendPreview = this._appendPreview;
 
   var actions = {};
 
   if (element.type === 'label') {
+    if (this._isDeleteAllowed([ element ])) {
+      addDeleteAction();
```
Comment (nikku) [inline lib/features/context-pad/ContextPadProvider.js:201]: Slightly less magic, and clear on the side-effects we cause:

```suggestion
      assign(actions, deleteAction());
```

Comment (nikku) [reply lib/features/context-pad/ContextPadProvider.js:201]: Or, alternatively:

```suggestion
      addDeleteAction(actions);
```

**Review by nikku — APPROVED** [review]
Great stuff. Thanks for following up on the review hints, too :cake:

**Pull request merged**